zebra-core/src/Zebra/Merge/Table.hs

```diff
@@ -96,156 +97,143 @@ peekHead input = do
      Monad m
-  => Input m
-  -> StateT (Map Logical.Value Int64) (EitherT UnionTableError m) (Input m)
-updateInput input =
-  case inputStream input of
-    Nothing ->
-      pure input
-    Just stream ->
-      if hasData input then
-        pure input
-      else do
-        e <- lift . lift $ Stream.next stream
-        case e of
-          Left () ->
-            pure $
-              closeStream input
-
-          Right (table, remaining) -> do
-            values <- lift . firstT UnionLogicalSchemaError . hoistEither $ Logical.takeMap table
-            modify' $ Map.unionWith (+) (Map.map Logical.sizeValue values)
-            pure $ Input values (Just remaining)
-{-# INLINABLE updateInput #-}
-
-takeExcessiveValues :: Maybe MaximumRowSize -> Map Logical.Value Int64 -> Map Logical.Value Int64
-takeExcessiveValues = \case
-  Nothing ->
-    const Map.empty
-  Just size ->
-    Map.filter (> unMaximumRowSize size)
-{-# INLINABLE takeExcessiveValues #-}
-
-unionStep :: Monad m => Logical.Value -> Cons Boxed.Vector (Input m) -> EitherT UnionTableError m (Step m)
-unionStep key inputs = do
-  step <- firstT UnionLogicalMergeError . hoistEither . (Logical.unionStep key) $ fmap inputData inputs
-  pure $
-    Step
-      (Logical.unionComplete step)
-      (Cons.zipWith replaceData (Logical.unionRemaining step) inputs)
-{-# INLINABLE unionStep #-}
-
-maximumKey :: Map Logical.Value Logical.Value -> Maybe Logical.Value
-maximumKey kvs =
-  if Map.null kvs then
-    Nothing
-  else
-    pure . fst $ Map.findMax kvs
-{-# INLINABLE maximumKey #-}
-
-unionInput ::
+  => Int
+  -> Stream (Of Striped.Table) m ()
+  -> Stream (Of Row) (EitherT UnionTableError m) ()
+streamStripedAsRows _num stream =
+  Stream.map (uncurry Row) $
+    Stream.concat $
+    Stream.mapM (hoistEither . logicalPairs) $
+    hoist lift
+      stream
+{-# INLINABLE streamStripedAsRows #-}
+
+mergeStreams ::
      Monad m
-  => Maybe MaximumRowSize
-  -> Cons Boxed.Vector (Input m)
-  -> Map Logical.Value Int64
-  -> Stream (Of Logical.Table) (EitherT UnionTableError m) ()
-unionInput msize inputs0 sizes0 = do
-  (inputs1, sizes1) <- lift $ runStateT (traverse updateInput inputs0) sizes0
-  unless (Cons.all isClosed inputs1) $ do
-    let
-      drops =
-        takeExcessiveValues msize sizes1
-
-      inputs2 =
-        fmap (dropData drops) inputs1
-
-      maximums =
-        Cons.mapMaybe (maximumKey . inputData) inputs1
-
-    if Boxed.null maximums then
-      unionInput msize inputs2 sizes1
-    else do
-      Step values inputs3 <- lift $ unionStep (Boxed.minimum maximums) inputs2
+  => Stream (Of Row) (EitherT UnionTableError m) ()
+  -> Stream (Of Row) (EitherT UnionTableError m) ()
+  -> Stream (Of Row) (EitherT UnionTableError m) ()
+mergeStreams leftStream rightStream = Effect . newEitherT $ do
+  left0  <- runEitherT $ Stream.next leftStream
+  right0 <- runEitherT $ Stream.next rightStream
+
+  return $ do
+    left1 <- left0
+    right1 <- right0
```

Comment:
These two are just monadic actions on `m`, which in our case, is `IO`. So this could be done with Async.